sphinx_js/typedoc.py

```diff
@@ -0,0 +1,299 @@
+# - optional access of [ public, private, protected ]
+# - optional classdesc
+# - optional comment (controls doclet inclusion)
+# - optional description
+# - optional exceptions
+# - optional exclude-members
+# -          kind of [ function, typedef, <other> ]
+# -          longname
+# - optional memberof
+# -          meta.filename
+# -          meta.lineno
+# -          meta.code.paramnames
+# -          meta.code.path
+# -          name
+# - optional params
+# - optional properties
+# - optional returns
+# -          type.names
+# - optional undocumented
+
+
+class Typedoc(object):
+
+    def get_parent(self, node):
+        parentId = node.get('__parentId')
+        return self.nodelist[parentId] if parentId is not None else None
+
+    def make_longname(self, node):
+        parent = self.get_parent(node)
+        longname = self.make_longname(parent) if parent is not None else ""
+
+        kindString = node.get('kindString')
+        if kindString in [None, 'Function', 'Constructor', 'Method']:
+            return longname
+        if longname != "":
+            flags = node.get('flags')
+            if (parent.get('kindString') in ['Class', 'Interface'] and
+                    flags.get('isStatic') is not True):
+                longname += '#'
+            elif parent.get('kindString') in ['Function', 'Method']:
+                longname += '.'
+            else:
+                longname += '~'
+        if kindString == 'Module':
+            return longname + "module:" + node.get("name")[1:-1]
+        elif kindString == 'External module':
+            return longname + "external:" + node.get("name")[1:-1]
+        else:
+            return longname + node.get("name")
+
+    def make_meta(self, node):
+        source = node.get('sources')[0]
+        return {
+            'path': os.path.dirname(source.get('fileName')) or './',
+            'filename': os.path.basename(source.get('fileName')),
+            'lineno': source.get('line'),
+            'code': {}
+        }
+
+    def extend_doclet(self, result, **kwargs):
+        for name in kwargs:
+            if kwargs[name] is None:
+                continue
+            result[name] = kwargs[name]
+        return result
+
+    def make_doclet(self, **kwargs):
+        return self.extend_doclet({}, **kwargs)
+
+    def make_type_name(self, type):
+        names = []
+        if type.get('type') == 'reference' and type.get('id'):
+            node = self.nodelist[type.get('id')]
+            # Should be: names = [ self.make_longname(node)]
+            parent = self.nodelist[node.get('__parentId')]
+            if parent.get('kindString') == 'External module':
+                names = [parent['name'][1:-1]+"."+node['name']]
+            else:
+                names = [node['name']]
+        elif type.get('type') in ['intrinsic', 'reference']:
+            names = [type.get('name')]
+        elif type.get('type') == 'stringLiteral':
+            names = ['"'+type.get('value')+'"']
```

Comment:
Rather than point all these little PEP 8 niggles out by hand, here's the result of a flake8 run:

% flake8 sphinx_js
sphinx_js/renderers.py:250:1: E302 expected 2 blank lines, found 1
sphinx_js/renderers.py:256:1: E302 expected 2 blank lines, found 1
sphinx_js/generators.py:17:31: E226 missing whitespace around arithmetic operator
sphinx_js/typedoc.py:84:46: E226 missing whitespace around arithmetic operator
sphinx_js/typedoc.py:84:50: E226 missing whitespace around arithmetic operator
sphinx_js/typedoc.py:90:25: E226 missing whitespace around arithmetic operator
sphinx_js/typedoc.py:90:43: E226 missing whitespace around arithmetic operator
sphinx_js/typedoc.py:92:66: E226 missing whitespace around arithmetic operator
sphinx_js/typedoc.py:121:21: E126 continuation line over-indented for hanging indent
sphinx_js/typedoc.py:123:17: E121 continuation line under-indented for hanging indent
sphinx_js/typedoc.py:177:37: E231 missing whitespace after ','
sphinx_js/typedoc.py:177:45: E225 missing whitespace around operator
sphinx_js/typedoc.py:179:1: W293 blank line contains whitespace
sphinx_js/typedoc.py:199:62: E226 missing whitespace around arithmetic operator
sphinx_js/typedoc.py:199:74: E226 missing whitespace around arithmetic operator

I've since merged the flake8 config that generated that into master, for when you want to re-run it yourself. Merging master back in would do you good anyway; it's been awhile.